Implement RemoteDesktop protocol

David96 opened the pull request:
Based on #263

Implements pointer and keyboard support, with most of the keyboard code copied from Wayvnc.

My usecase ist the kdeconnect remote input, so that's how it is tested. Sometimes sway pointer input seems to be completely broken after a while, see ce01c224 for a try to fix that (not sure whether this is the right approach).
Keyboard input works mostly but somehow capital letters don't, not sure whether this is a kdeconnect bug or in this implementation.

It also implements ScreenCast via the RemoteDesktop protocol, so Zoom screen sharing works again.

David96 commented:
For some reason the BSD build has "SHM_ANON" not defined although it's supposed to be supported by FreeBSD. Not sure what to do about that.

David96 commented:
!!! Last commit crashes your compositor without https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4980 when scrolling !!!

Bot-wxt1221 commented:
It makes rustdesk work! However, mouse is still not available. Anyway, It works with kdeconnect. I'm not sure what's wrong.

Bot-wxt1221 commented:
@David96 niri implent wlr-virtual-pointer https://github.com/YaLTeR/niri/pull/630

David96 commented:
Ah I see. But it doesn't seem like rustdesk supports Wayland mouse input yet: https://github.com/rustdesk/rustdesk/tree/master/libs/enigo - therefore kdeconnect working and rustdesk not working is expected behaviour and has to be fixed by rustdesk.

Bot-wxt1221 commented:
@David96 I'm not sure. But it is seemed that only README is not edited. https://github.com/rustdesk/rustdesk/commit/00867276eda18923805ef78e15c341420a3dd04d

David96 commented on src/remotedesktop/remotedesktop.c:
```diff
@@ -0,0 +1,892 @@
+	ret = sd_bus_message_skip(msg, "a{sv}");
+	if (ret < 0) {
+		return ret;
+	}
```
A log might be good here, that's true but I don't think the code is wrong?
If the message doesn't include this element it's not a valid request as per the protocol. At least my understanding is that while the options element includes optional configuration, it's not optional to include the element itself (even if empty) in the message.

Bot-wxt1221 commented:
Also, we can implent EIS like https://github.com/gac3k/xdg-desktop-portal-hypr-remote. However, I'm not interested in EIS though.

alimg commented:
> !!! Last commit crashes your compositor without https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4980 when scrolling !!!


Since there is no progress on your merge request for the last 6 months, I did a ugly workaround for my personal use :upside_down_face: 

So far scrolling works with this https://github.com/alimg/xdg-desktop-portal-wlr/commit/

Bot-wxt1221 commented on src/remotedesktop/remotedesktop.c:
```diff
@@ -0,0 +1,892 @@
+	zwlr_virtual_pointer_v1_motion_absolute(sess->remotedesktop_data.virtual_pointer,
+		get_timestamp_ms(&sess->remotedesktop_data),
+		wl_fixed_from_double(x), wl_fixed_from_double(y),
+		output->width, output->height);
```
It just work. I think all argument in int or double should be passed by wl_fixed_xxx

David96 committed:
remotedesktop: add keyboard code from wavnc

The files are mostly taken verbatim from https://github.com/any1/wayvnc after
commit 9e85093a6dc164da3c605b27336a0197252e688d.

Modifications made to work with XDPW:
- Remove neatvnc dependency
- Add timing information to sent keycodes
- Use existing XDPW shm code

David96 committed:
remotedesktop: add keyboard support

David96 committed:
remotedesktop: Add screencast support

emersion commented on src/remotedesktop/shm.c:
We already have code in xdpw to allocate shared memory buffers, in `src/screencast/screencast_common.c`. The shared bits can be move in a common location, e.g. `src/shm.c`.

emersion replied:
In general we prefer to stick to POSIX APIs.

David96 commented:
Thanks for the feedback! Using the recipe style was my idea for the latest rebase. Since the first two commits are not made by me, I did not want to split them up however - which is the reason I left them as is and added a cleanup commit afterwards.

I'm not sure what the correct procedure would be here regarding author ship - I guess I can split them up anyways, but keep the original authorship and incorporate my cleanups? Is that fine with you, @andreafeletto ?

Commit:
remotedesktop: Implement DBus interface

Co-authored-by: David96 <david@hameipe.de>

Commit:
remotedesktop: allow for device selection

This adds configuration options to allow for pointer/keyboard
control and implements device selection via the DBus protocol.

Co-authored-by: David96 <david@hameipe.de>

Commit:
remotedestop: Implement pointer input

Co-authored-by: David96 <david@hameipe.de>

emersion commented on src/remotedesktop/remotedesktop.c:
```diff
@@ -235,32 +250,325 @@ static int method_remotedesktop_start(sd_bus_message *msg, void *data, sd_bus_er
+
+	if (btn_state == WL_POINTER_BUTTON_STATE_PRESSED) {
+		if (sess->remotedesktop_data.pressed_buttons & 1<<button) {
+			logprint(WARN, "remotedesktop: npb: pointer already pressed, releasing first");
```
Is that what the portal spec says we should be doing? It sounds a bit weird to me to send release implicitly.

emersion commented on src/core/shm_util.c:
```diff
@@ -0,0 +1,57 @@
+#include "shm_util.h"
```
Style: local includes should come last in the list

emersion commented on src/core/shm_util.c:
Can the move of the shm code be in its own separate commit?

emersion commented on include/intset.h:
Yeah, sorry, this comment is about the entirety of the wayvnc imported code. I didn't find a better place to post it.

David96 committed:
screencast_common: split of SHM code into shm_util.c

David96 commented on src/remotedesktop/remotedesktop.c:
```diff
@@ -235,32 +250,325 @@ static int method_remotedesktop_start(sd_bus_message *msg, void *data, sd_bus_er
+	if (!(sess->remotedesktop_data.devices & POINTER)) {
+		logprint(ERROR, "remotedesktop: npma: called, but pointer not selected!");
+		return -1;
+	}
```
I noticed that I already put the for_each loop into its own method. Would you still like to have the whole bunch abstracted or do you think that is sufficient?

David96 commented on src/core/config.c:
```diff
@@ -85,13 +86,37 @@ static int handle_ini_screencast(struct config_screencast *screencast_conf, cons
+		logprint(TRACE, "config: skipping invalid key in config file");
+		return 0;
+	}
+	if (is_allowed) {
```
Should be fixed now - thanks for the report!

It actually made me find a typo I made in the definition of the default config: I set the default to `KEYBOARD & POINTER` instead of `KEYBOARD | POINTER` :facepalm:

shimeike commented:
> My usecase ist the kdeconnect remote input,

This is my desired use case also. Is there anything specific that needs to be done/configured for this to work once this is merged?

David96 commented:
It should just work. If you want to test it before it is merged you'll have to make sure that the remotedesktop interface is included in the wlr.portal file (the system wide is located at /usr/share/xdg-desktop-portal/portals there's probably a user overwrite for it but I don't remember from the top of my head).

shimeike commented:
> It should just work. If you want to test it before it is merged you'll have to make sure that the remotedesktop interface is included in the wlr.portal file (the system wide is located at /usr/share/xdg-desktop-portal/portals there's probably a user overwrite for it but I don't remember from the top of my head).

Thanks for the advice - it works! :smiley:
